### ISSUE TITLE
Filter qubits from the circuit  and a mutable circuit width

### ISSUE THREAD

**Issue by Cryoris** [Qiskit/qiskit-terra#3995]
### What is the expected behavior?

Add a feature to change the number of qubits after the circuit construction. If the number of qubits increases, add more qubits -- as already possible by adding registers. If the number of qubits decreases remove the bottom qubits and all operations associated with it. This could be generalized to a filter method removing specific qubits from the circuit and operations they are part of.

Note that a filter feature can be arbitrarily generalized if we feel the need for it. E.g. we could filter certain gates, certain parameters, etc. Or a filter could also be used to search for certain operations. **However** the filter functionality is **not required** to move along with changes in Aqua, just reducing the number of qubits is. The filter is merely a suggestion of how to embed the reduction in a nice framework.

Use cases
1. A mutable circuit width is required generally to make circuits work with Aqua algorithms and remove the layer of wrapping circuits into Aqua-compatible objects 
2. Concrete use case: It is possible to compute the gradient of a circuit by adding a gate controlled by an ancilla qubit. The original circuit can be retrieved by removing the ancilla and the operations associated with it.

```
>>> qc = QuantumCircuit(2)
>>> qc.x(0)
>>> qc.cry(0.2, 0, 1)
>>> qc.draw()
        ┌───┐           
q_0: |0>┤ X ├─────■─────
        └───┘┌────┴────┐
q_1: |0>─────┤ Ry(0.2) ├
             └─────────┘
>>> qc.num_qubits = 1
>>> qc.draw()
        ┌───┐
q_0: |0>┤ X ├
        └───┘
>>> qc.num_qubits = 4
>>> qc.draw()
        ┌───┐
q_0: |0>┤ X ├
        └───┘
q_1: |0>─────
             
q_2: |0>─────
             
q_3: |0>─────
             
```

Or using the filter functionality
```
# qc = ...
>>> qc.draw()
                       ┌───┐
q_0: |0>──■────────────┤ X ├
        ┌─┴─┐          └─┬─┘
q_1: |0>┤ X ├──■─────────┼──
        └───┘┌─┴─┐       │  
q_2: |0>─────┤ X ├──■────┼──
             └───┘┌─┴─┐  │  
q_3: |0>──────────┤ X ├──■──
                  └───┘     
>>> qc.filter(qubit=2)
>>> qc.draw()
             ┌───┐
q_0: |0>──■──┤ X ├
        ┌─┴─┐└─┬─┘
q_1: |0>┤ X ├──┼──
        └───┘  │  
q_3: |0>───────■──
```

**Comment by BKGodd** [comment]
Hello, I have wanted to implement some functionality similar to this myself, as well. I would love to be able to contribute to this if no one is opposed.

**Comment by fs1132429** [comment]
### Update for GoGP
This issue may be fixed by PR #5033 which is pending review.

**Comment by jakelishman** [comment]
This was resolved by #4185 for the use cases that we're prepared to support.
